src/ReplicatedStorage/Common/Util/StatefulIncrementor.lua

```diff
@@ -0,0 +1,191 @@
+local ReplicatedStorage = game:GetService("ReplicatedStorage")
+local TweenService = game:GetService("TweenService")
+local Silo = require(ReplicatedStorage.Packages.Silo)
+
+--[=[
+	The StatefulIncrementor passively increments a value over a state, without spawning extra threads or timers.
+	An incrementation may be started, then collapsed or queried later.
+
+	@class StatefulIncrementor
+]=]
+local StatefulIncrementor = {}
+StatefulIncrementor.__index = StatefulIncrementor
+
+--[=[
+	Returns the current time.
+	May be overriden per-incrementor or globally.
+
+	@return number
+]=]
+function StatefulIncrementor.getTime()
+	return workspace:GetServerTimeNow()
+end
+
+--[=[
+	@return boolean -- Whether or not incrementation is in progress.
+]=]
+function StatefulIncrementor:IsIncrementing(): boolean
+	return not not self.Time
+end
+
+--[=[
+	@return boolean -- Whether or not incrementation has expired.
+]=]
+function StatefulIncrementor:IsExpired(): boolean
+	if not self:IsIncrementing() then
+		return false
+	end
+	return self:GetProgress() >= 1
+end
+
+--[=[
+	Returns the raw progress of the incrementor as a value from 0 -> 1
+	@param duration number? -- The duration to calculate over.
+	@return number
+]=]
+function StatefulIncrementor:GetRawProgress(duration: number?): number
+	if not self:IsIncrementing() then
+		return 0
+	end
+	local timeElapsed = self.getTime() - self.Time
```

Comment:
self.Time is a very ambiguously named variable. I wasn't really sure what it was until I read the entire code

Reply:
I'm renaming this to `self.StartTime`